server/src/main/java/io/druid/server/coordinator/helper/DruidCoordinatorLogger.java

```diff
@@ -213,91 +197,92 @@ public Number transformEntry(String key, AtomicLong value)
+          emitter.emit(
               new ServiceMetricEvent.Builder()
-                      .setDimension(DruidMetrics.DATASOURCE, datasource).build(
-                      "segment/unavailable/count", count
+                  .setDimension(DruidMetrics.DATASOURCE, dataSource).build(
+                  "segment/unavailable/count", count
               )
-      );
-    }
-    for (Map.Entry<String, CountingMap<String>> entry : coordinator.getReplicationStatus().entrySet()) {
-      String tier = entry.getKey();
-      CountingMap<String> datasourceAvailabilities = entry.getValue();
-      for (Map.Entry<String, AtomicLong> datasourceAvailability : datasourceAvailabilities.entrySet()) {
-        String datasource = datasourceAvailability.getKey();
-        Long count = datasourceAvailability.getValue().get();
-        emitter.emit(
+          );
+        }
+    );
+
+    coordinator.getReplicationStatus().forEach(
+        (final String tier, final Object2LongMap<String> status) -> {
+          for (final Object2LongMap.Entry<String> entry : status.object2LongEntrySet()) {
+            final String dataSource = entry.getKey();
+            final long count = entry.getLongValue();
+
+            emitter.emit(
                 new ServiceMetricEvent.Builder()
-                        .setDimension(DruidMetrics.TIER, tier)
-                        .setDimension(DruidMetrics.DATASOURCE, datasource).build(
-                        "segment/underReplicated/count", count
+                    .setDimension(DruidMetrics.TIER, tier)
+                    .setDimension(DruidMetrics.DATASOURCE, dataSource).build(
+                    "segment/underReplicated/count", count
                 )
-        );
-      }
-    }
+            );
+          }
+        }
+    );
 
     // Emit segment metrics
-    CountingMap<String> segmentSizes = new CountingMap<String>();
-    CountingMap<String> segmentCounts = new CountingMap<String>();
-    for (DruidDataSource dataSource : params.getDataSources()) {
-      for (DataSegment segment : dataSource.getSegments()) {
-        segmentSizes.add(dataSource.getName(), segment.getSize());
-        segmentCounts.add(dataSource.getName(), 1L);
-      }
-    }
-    for (Map.Entry<String, Long> entry : segmentSizes.snapshot().entrySet()) {
-      String dataSource = entry.getKey();
-      Long size = entry.getValue();
-      emitter.emit(
-          new ServiceMetricEvent.Builder()
-              .setDimension(DruidMetrics.DATASOURCE, dataSource).build(
-              "segment/size", size
-          )
-      );
-    }
-    for (Map.Entry<String, Long> entry : segmentCounts.snapshot().entrySet()) {
-      String dataSource = entry.getKey();
-      Long count = entry.getValue();
-      emitter.emit(
-          new ServiceMetricEvent.Builder()
-              .setDimension(DruidMetrics.DATASOURCE, dataSource).build(
-              "segment/count", count
-          )
-      );
-    }
+    final Stream<DataSegment> allSegments = params
```

Comment:
Not sure why here segments are first merged from a per-data source grouping into a single Stream, and then grouped by data source again. I'll fix this.